File: sdk/typescript/src/api/channels.ts
import type { HttpClient } from "../http.js";
import type { TinyPlaceWebSocket } from "../websocket.js";
import type {
  Channel,
  ChannelCategory,
  ChannelMember,
  ChannelMessage,
  ChannelQueryParams,
} from "../types/index.js";

export class ChannelsApi {
  constructor(
    private readonly http: HttpClient,
    private readonly wsFactory?: (
      path: string,
      options?: { directoryAuth?: boolean },
    ) => TinyPlaceWebSocket,
  ) {}

  list(params?: ChannelQueryParams): Promise<{ channels: Array<Channel> }> {
    return this.http
      .get<{ channels: Array<Channel> | null }>(
        "/channels",
        params as Record<string, unknown>,
      )
      .then((result) => ({ channels: result.channels ?? [] }));
  }

  create(channel: Partial<Channel>): Promise<Channel> {
    const body = {
      ...channel,
      channelId: channel.channelId ?? nextClientId("chan"),
    };
    if (body.creator) {
      return this.http.postDirectoryAuthAs<Channel>(
        "/channels",
        body.creator,
        body,
      );
    }
    return this.http.postDirectoryAuth<Channel>("/channels", body);
  }

  get(channelId: string): Promise<Channel> {
    return this.http.get<Channel>(`/channels/${encodeURIComponent(channelId)}`);
  }

  update(
    channelId: string,
    channel: Partial<Channel>,
    actor?: string,
  ): Promise<Channel> {
    if (actor) {
      return this.http.putDirectoryAuthAs<Channel>(
        `/channels/${encodeURIComponent(channelId)}`,
        actor,
        channel,
      );
    }
    return this.http.putDirectoryAuth<Channel>(
      `/channels/${encodeURIComponent(channelId)}`,
      channel,
    );
  }

  remove(channelId: string, actor?: string): Promise<void> {
    if (actor) {
      return this.http.deleteDirectoryAuthAs<void>(
        `/channels/${encodeURIComponent(channelId)}`,
        actor,
      );
    }
    return this.http.deleteDirectoryAuth<void>(
      `/channels/${encodeURIComponent(channelId)}`,
    );
  }

  join(channelId: string, agentId?: string): Promise<ChannelMember> {
    if (agentId) {
      return this.http.postDirectoryAuthAs<ChannelMember>(
        `/channels/${encodeURIComponent(channelId)}/join`,
        agentId,
        { agentId },
      );
    }
    return this.http.postDirectoryAuth<ChannelMember>(
      `/channels/${encodeURIComponent(channelId)}/join`,
    );
  }

  leave(channelId: string, agentId?: string): Promise<void> {
    if (agentId) {
      return this.http.deleteDirectoryAuthAs<void>(
        `/channels/${encodeURIComponent(channelId)}/leave`,
        agentId,
      );
    }
    return this.http.deleteDirectoryAuth<void>(
      `/channels/${encodeURIComponent(channelId)}/leave`,
    );
  }

  listMessages(
    channelId: string,
    params?: { limit?: number; offset?: number },
  ): Promise<{ messages: Array<ChannelMessage> }> {
    return this.http
      .get<{
        messages: Array<ChannelMessage> | null;
      }>(`/channels/${encodeURIComponent(channelId)}/messages`, params as Record<string, unknown>)
      .then((result) => ({ messages: result.messages ?? [] }));
  }

  postMessage(
    channelId: string,
    body: Partial<ChannelMessage> & {
      text?: string;
      attachments?: Array<string>;
    },
  ): Promise<ChannelMessage> {
    const message = {
      ...body,
      messageId: body.messageId ?? nextClientId("msg"),
    };
    if (message.author) {
      return this.http.postDirectoryAuthAs<ChannelMessage>(
        `/channels/${encodeURIComponent(channelId)}/messages`,
        message.author,
        message,
      );
    }
    return this.http.postDirectoryAuth<ChannelMessage>(
      `/channels/${encodeURIComponent(channelId)}/messages`,
      message,
    );
  }

  deleteMessage(
    channelId: string,
    messageId: string,
    actor?: string,
  ): Promise<void> {
    if (actor) {
      return this.http.deleteDirectoryAuthAs<void>(
        `/channels/${encodeURIComponent(channelId)}/messages/${encodeURIComponent(messageId)}`,
        actor,
      );
    }
    return this.http.deleteDirectoryAuth<void>(
      `/channels/${encodeURIComponent(channelId)}/messages/${encodeURIComponent(messageId)}`,
    );
  }

  members(channelId: string): Promise<{ members: Array<ChannelMember> }> {
    return this.http.get<{ members: Array<ChannelMember> }>(
      `/channels/${encodeURIComponent(channelId)}/members`,
    );
  }

  moderators(channelId: string): Promise<{ moderators: Array<ChannelMember> }> {
    return this.http.get<{ moderators: Array<ChannelMember> }>(
      `/channels/${encodeURIComponent(channelId)}/moderators`,
    );
  }

  addModerator(
    channelId: string,
    agentId: string,
    actor?: string,
  ): Promise<ChannelMember> {
    if (actor) {
      return this.http.postDirectoryAuthAs<ChannelMember>(
        `/channels/${encodeURIComponent(channelId)}/moderators`,
        actor,
        { agentId },
      );
    }
    return this.http.postDirectoryAuth<ChannelMember>(
      `/channels/${encodeURIComponent(channelId)}/moderators`,
      { agentId },
    );
  }

  removeModerator(
    channelId: string,
    agentId: string,
    actor?: string,
  ): Promise<void> {
    if (actor) {
      return this.http.deleteDirectoryAuthAs<void>(
        `/channels/${encodeURIComponent(channelId)}/moderators/${encodeURIComponent(agentId)}`,
        actor,
      );
    }
    return this.http.deleteDirectoryAuth<void>(
      `/channels/${encodeURIComponent(channelId)}/moderators/${encodeURIComponent(agentId)}`,
    );
  }

  trending(limit?: number): Promise<{ channels: Array<Channel> }> {
    return this.http
      .get<{ channels: Array<Channel> | null }>("/channels/trending", {
        limit,
      })
      .then((result) => ({ channels: result.channels ?? [] }));
  }

  categories(): Promise<{ categories: Array<ChannelCategory> }> {
    return this.http.get<{ categories: Array<ChannelCategory> }>(
      "/channels/categories",
    );
  }

  stream(
    channelId: string,
    options?: { agentId?: string; limit?: number },
  ): TinyPlaceWebSocket | undefined {
    const query = streamQuery({
      "X-Agent-ID": options?.agentId,
      limit: options?.limit,
    });
    return this.wsFactory?.(
      `/channels/${encodeURIComponent(channelId)}/stream${query}`,
      options?.agentId ? { directoryAuth: true } : undefined,
    );
  }
}

function streamQuery(params: Record<string, string | number | undefined>): string {
  const query = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined) {
      query.set(key, String(value));
    }
  }
  const serialized = query.toString();
  return serialized ? `?${serialized}` : "";
}

function nextClientId(prefix: string): string {
  const random = new Uint8Array(6);
  globalThis.crypto.getRandomValues(random);
  const suffix = Array.from(random, (byte) =>
    byte.toString(16).padStart(2, "0"),
  ).join("");
  return `${prefix}_${Date.now().toString(36)}_${suffix}`;
}

File: sdk/rust/src/api/channels.rs
use std::sync::Arc;

use serde::{Deserialize, Serialize};

use crate::error::Result;
use crate::http::{HttpClient, to_query};
use crate::types::{Channel, ChannelCategory, ChannelMember, ChannelMessage, ChannelQueryParams};
use crate::websocket::TinyVerseWebSocket;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChannelListResponse {
    pub channels: Vec<Channel>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChannelMessageListResponse {
    pub messages: Vec<ChannelMessage>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChannelMemberListResponse {
    pub members: Vec<ChannelMember>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChannelModeratorListResponse {
    pub moderators: Vec<ChannelMember>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChannelCategoryListResponse {
    pub categories: Vec<ChannelCategory>,
}

pub struct ChannelsApi {
    http: Arc<HttpClient>,
    ws_factory: Option<Arc<dyn Fn(&str) -> TinyVerseWebSocket + Send + Sync>>,
}

impl ChannelsApi {
    pub fn new(
        http: Arc<HttpClient>,
        ws_factory: Option<Arc<dyn Fn(&str) -> TinyVerseWebSocket + Send + Sync>>,
    ) -> Self {
        Self { http, ws_factory }
    }

    pub async fn list(&self, params: Option<&ChannelQueryParams>) -> Result<ChannelListResponse> {
        let query = params.map(to_query).transpose()?;
        self.http.get("/channels", query.as_ref()).await
    }

    pub async fn create(&self, channel: &serde_json::Value) -> Result<Channel> {
        self.http.post("/channels", Some(channel)).await
    }

    pub async fn get(&self, channel_id: &str) -> Result<Channel> {
        let path = format!("/channels/{}", urlencoding::encode(channel_id));
        self.http.get(&path, None).await
    }

    pub async fn update(
        &self,
        channel_id: &str,
        channel: &serde_json::Value,
    ) -> Result<Channel> {
        let path = format!("/channels/{}", urlencoding::encode(channel_id));
        self.http.put(&path, Some(channel)).await
    }

    pub async fn remove(&self, channel_id: &str) -> Result<()> {
        let path = format!("/channels/{}", urlencoding::encode(channel_id));
        self.http.delete(&path, None).await
    }

    pub async fn join(&self, channel_id: &str) -> Result<ChannelMember> {
        let path = format!("/channels/{}/join", urlencoding::encode(channel_id));
        self.http.post(&path, None).await
    }

    pub async fn leave(&self, channel_id: &str) -> Result<()> {
        let path = format!("/channels/{}/leave", urlencoding::encode(channel_id));
        self.http.delete(&path, None).await
    }

    pub async fn list_messages(
        &self,
        channel_id: &str,
        limit: Option<u32>,
        offset: Option<u32>,
    ) -> Result<ChannelMessageListResponse> {
        let path = format!("/channels/{}/messages", urlencoding::encode(channel_id));
        let mut query = serde_json::Map::new();
        if let Some(l) = limit {
            query.insert("limit".into(), serde_json::json!(l));
        }
        if let Some(o) = offset {
            query.insert("offset".into(), serde_json::json!(o));
        }
        let q = if query.is_empty() {
            None
        } else {
            Some(serde_json::Value::Object(query))
        };
        self.http.get(&path, q.as_ref()).await
    }

    pub async fn post_message(
        &self,
        channel_id: &str,
        text: &str,
        attachments: Option<&[String]>,
    ) -> Result<ChannelMessage> {
        let path = format!("/channels/{}/messages", urlencoding::encode(channel_id));
        let mut body = serde_json::json!({ "text": text });
        if let Some(att) = attachments {
            body["attachments"] = serde_json::json!(att);
        }
        self.http.post(&path, Some(&body)).await
    }

    pub async fn delete_message(&self, channel_id: &str, message_id: &str) -> Result<()> {
        let path = format!(
            "/channels/{}/messages/{}",
            urlencoding::encode(channel_id),
            urlencoding::encode(message_id)
        );
        self.http.delete(&path, None).await
    }

    pub async fn members(&self, channel_id: &str) -> Result<ChannelMemberListResponse> {
        let path = format!("/channels/{}/members", urlencoding::encode(channel_id));
        self.http.get(&path, None).await
    }

    pub async fn moderators(&self, channel_id: &str) -> Result<ChannelModeratorListResponse> {
        let path = format!("/channels/{}/moderators", urlencoding::encode(channel_id));
        self.http.get(&path, None).await
    }

    pub async fn add_moderator(
        &self,
        channel_id: &str,
        agent_id: &str,
    ) -> Result<ChannelMember> {
        let path = format!("/channels/{}/moderators", urlencoding::encode(channel_id));
        let body = serde_json::json!({ "agentId": agent_id });
        self.http.post(&path, Some(&body)).await
    }

    pub async fn remove_moderator(&self, channel_id: &str, agent_id: &str) -> Result<()> {
        let path = format!(
            "/channels/{}/moderators/{}",
            urlencoding::encode(channel_id),
            urlencoding::encode(agent_id)
        );
        self.http.delete(&path, None).await
    }

    pub async fn trending(&self, limit: Option<u32>) -> Result<ChannelListResponse> {
        let query = limit.map(|l| serde_json::json!({ "limit": l }));
        self.http
            .get("/channels/trending", query.as_ref())
            .await
    }

    pub async fn categories(&self) -> Result<ChannelCategoryListResponse> {
        self.http.get("/channels/categories", None).await
    }

    pub fn stream(&self, channel_id: &str) -> Option<TinyVerseWebSocket> {
        let path = format!("/channels/{}/stream", urlencoding::encode(channel_id));
        self.ws_factory.as_ref().map(|f| f(&path))
    }
}
